sentry/src/main/java/io/sentry/SentryClient.java

```diff
@@ -582,14 +582,29 @@ public void captureSession(final @NotNull Session session, final @Nullable Hint
 
     try {
       hint.clear();
-      transport.send(envelope, hint);
+      return sendEnvelope(envelope, hint);
     } catch (IOException e) {
       options.getLogger().log(SentryLevel.ERROR, "Failed to capture envelope.", e);
-      return SentryId.EMPTY_ID;
     }
-    final SentryId eventId = envelope.getHeader().getEventId();
-    if (eventId != null) {
-      return eventId;
+    return SentryId.EMPTY_ID;
+  }
+
+  private @NotNull SentryId sendEnvelope(
+      @NotNull final SentryEnvelope envelope, @Nullable final Hint hint) throws IOException {
+    final @Nullable SentryOptions.BeforeEnvelopeCallback beforeEnvelopeCallback =
+        options.getBeforeEnvelopeCallback();
+    @Nullable SentryEnvelope envelopeToSend = envelope;
+    if (beforeEnvelopeCallback != null) {
+      envelopeToSend = beforeEnvelopeCallback.execute(envelope, hint);
+    }
+    if (envelopeToSend != null) {
+      if (hint == null) {
+        transport.send(envelope);
+      } else {
+        transport.send(envelope, hint);
+      }
+      final @Nullable SentryId id = envelope.getHeader().getEventId();
+      return id != null ? id : SentryId.EMPTY_ID;
     } else {
       return SentryId.EMPTY_ID;
```

Comment:
in theory we'd have to record a client report on the lost envelope here with `DiscardReason.BEFORE_SEND`, but dunno if other platforms cared about that

Reply:
Good point, I guess it's fine for now - as beforeEnvelope is internal only. We could also change beforeEnvelope to not allow any filtering. At least e.g. RN does not allow that: https://github.com/getsentry/sentry-react-native/blob/a5f6b2f078f0434fcd46545b6b4f789d58b30769/src/js/client.ts#L138-L149

Reply:
yeah I'm good with whatever other sdks do in this case 👍

Reply:
I've switched to a `void` return type for now. less code, less complexity 😅